File: leetcode/python/215_Kth_Largest_Element_in_an_Array.py
# https://leetcode.com/problems/kth-largest-element-in-an-array/

class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        min_heap = MinHeap(k)
        for num in nums:
            if num > min_heap.get_first_value():
                min_heap.replace(num)
            # min_heap.print_value()
        return min_heap.get_first_value()
        
class MinHeap(object):
    def __init__(self, k):
        """
        :type k: int
        :rtype: List[int]
        """
        self.min_heap = [-sys.maxint] * k
    
    def get_first_value(self):
        """
        rtype: int
        """
        return self.min_heap[0]
        
    def replace(self, n):
        """
        :type n: int
        """
        self.min_heap[0] = n
        self.reorganize(0)    
    
    def reorganize(self, n):
        """
        :type n: int
        """
        if n < len(self.min_heap):
            current_value = self.min_heap[n]
        else:
            return True
        left_child_idx = 2 * n + 1
        right_child_idx = 2 * n + 2
        if left_child_idx < len(self.min_heap):
            left_child_value = self.min_heap[left_child_idx]
        else:
            left_child_value = sys.maxint
        if right_child_idx < len(self.min_heap):
            right_child_value = self.min_heap[right_child_idx]
        else:
            right_child_value = sys.maxint
        if left_child_value < right_child_value:
            pos = left_child_idx
            smaller_child_value = left_child_value
        else:
            pos = right_child_idx
            smaller_child_value = right_child_value
        if current_value > smaller_child_value:
            self.min_heap[n] = smaller_child_value
            self.min_heap[pos] = current_value
            self.reorganize(pos)
        else:
            return True
        return True
    
    def print_value(self):
        """
        """
        print(self.min_heap)

File: leetcode/c++/215_Kth_Largest_Element_in_an_Array.cpp
// https://leetcode.com/problems/kth-largest-element-in-an-array/

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        MinHeap min_heap (k);
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] > min_heap.getFirstValue()) {
                min_heap.replace(nums[i]);
            }
            // min_heap.print();
        }
        return min_heap.getFirstValue();
    }
    
    class MinHeap {
    public:
        MinHeap(int k) {
            this->min_heap.resize(k, INT_MIN);
        }
        
        // ~MinHeap(); // adding this gives a compiling error!!!
        
        int getFirstValue() {
            return this->min_heap[0];
        }
        
        bool replace(int num) {
            this->min_heap[0] = num;
            this->reorganize(0);
            return true;
        }
        
        bool reorganize(int idx) {
            if (idx >= this->min_heap.size()) {
                return true;
            }
            int current_value = this->min_heap[idx];
            int left_child_idx = idx * 2 + 1;
            int right_child_idx = idx * 2 + 2;
            int left_child_value = INT_MAX;
            int right_child_value = INT_MAX;
            if (left_child_idx < this->min_heap.size()) {
                left_child_value = this->min_heap[left_child_idx];
            }
            if (right_child_idx < this->min_heap.size()) {
                right_child_value = this->min_heap[right_child_idx];
            }
            int min_child_idx = -1;
            int min_child_value = INT_MAX;
            if (left_child_value < right_child_value) {
                min_child_value = left_child_value;
                min_child_idx = left_child_idx;
            } else {
                min_child_value = right_child_value;
                min_child_idx = right_child_idx;
            }
            if (current_value > min_child_value) {
                this->min_heap[idx] = min_child_value;
                this->min_heap[min_child_idx] = current_value;
                this->reorganize(min_child_idx);
            }
            return true;
        }
        
        void print() {
            for (int i = 0; i < this->min_heap.size(); ++i) {
                cout << this->min_heap[i] << " ";
            }
            cout << endl;
        }
    private:
        vector<int> min_heap;
    };
};
